rust/kernel/init.rs

```diff
@@ -0,0 +1,827 @@
+        unsafe { init.__pinned_init(slot)? };
+        // SAFETY: all fields have been initialized
+        Ok(unsafe { Pin::new_unchecked(this.assume_init().into()) })
+    }
+
+    fn init<E>(init: impl Init<T, E>) -> error::Result<Self>
+    where
+        Error: From<E>,
+    {
+        let mut this = UniqueArc::try_new_uninit()?;
+        let slot = this.as_mut_ptr();
+        // SAFETY: when init errors/panics, slot will get deallocated but not dropped,
+        // slot is valid
+        unsafe { init.__init(slot)? };
+        // SAFETY: all fields have been initialized
+        Ok(unsafe { this.assume_init().into() })
+    }
+}*/
+
+impl<T> InPlaceInit<T> for UniqueArc<T> {
+    fn pin_init<E>(init: impl PinInit<T, E>) -> error::Result<Pin<Self>>
+    where
+        Error: From<E>,
+    {
+        let mut this = UniqueArc::try_new_uninit()?;
+        let slot = this.as_mut_ptr();
+        // SAFETY: when init errors/panics, slot will get deallocated but not dropped,
+        // slot is valid and will not be moved because of the `Pin::new_unchecked`
+        unsafe { init.__pinned_init(slot)? };
+        // SAFETY: all fields have been initialized
+        Ok(unsafe { Pin::new_unchecked(this.assume_init()) })
+    }
+
+    fn init<E>(init: impl Init<T, E>) -> error::Result<Self>
+    where
+        Error: From<E>,
+    {
+        let mut this = UniqueArc::try_new_uninit()?;
+        let slot = this.as_mut_ptr();
+        // SAFETY: when init errors/panics, slot will get deallocated but not dropped,
+        // slot is valid
+        unsafe { init.__init(slot)? };
+        // SAFETY: all fields have been initialized
+        Ok(unsafe { this.assume_init() })
+    }
+}
+
+/// Marker trait for types that can be initialized by writing just zeroes.
+///
+/// # Safety
+/// The bit pattern consisting of only zeroes must be a valid bit pattern for the type.
+pub unsafe trait Zeroable {}
+
+/// Create a new zeroed T
+pub fn zeroed<T: Zeroable + Unpin>() -> impl Init<T> {
+    unsafe {
+        InitClosure::from_closure(|slot: *mut T| {
+            slot.write_bytes(0, 1);
+            Ok(())
+        })
+    }
+}
+
+/// An initializer that leaves the memory uninitialized.
+pub fn uninit<T>() -> impl Init<MaybeUninit<T>> {
+    unsafe { InitClosure::from_closure(|_| Ok(())) }
+}
+
+/// Convert a value into an initializer
+pub fn from_value<T>(value: T) -> impl Init<T> {
+    unsafe {
+        InitClosure::from_closure(move |slot: *mut T| {
+            slot.write(value);
+            Ok(())
+        })
+    }
+}
+
+macro_rules! impl_zeroable {
+    ($($t:ty),*) => {
+        $(unsafe impl Zeroable for $t {})*
+    };
+}
+// all primitives that are allowed to be 0
+impl_zeroable!(
+    bool, char, u8, u16, u32, u64, u128, usize, i8, i16, i32, i64, i128, isize, f32, f64
+);
+// there is nothing to zero
+impl_zeroable!(core::marker::PhantomPinned, Infallible, ());
+
+// we are allowed to zero padding bytes
```

Comment:
```suggestion
// We are allowed to zero padding bytes.
```

(and similarly elsewhere)